File: frontend/src/lib/api/notificationChannels.ts
import { api } from './client';
import type {
	NotificationChannelRead,
	NotificationChannelCreate,
	NotificationChannelUpdate
} from '$lib/types/notification-channel';

export const notificationChannelsApi = {
	list: (): Promise<NotificationChannelRead[]> => {
		return api.get<NotificationChannelRead[]>('/notification-channels');
	},

	get: (id: string): Promise<NotificationChannelRead> => {
		return api.get<NotificationChannelRead>(`/notification-channels/${id}`);
	},

	create: (data: NotificationChannelCreate): Promise<NotificationChannelRead> => {
		return api.post<NotificationChannelRead>('/notification-channels', data);
	},

	update: (id: string, data: NotificationChannelUpdate): Promise<NotificationChannelRead> => {
		return api.patch<NotificationChannelRead>(`/notification-channels/${id}`, data);
	},

	remove: (id: string): Promise<void> => {
		return api.delete(`/notification-channels/${id}`);
	},

	test: (id: string): Promise<{ success: boolean; message: string }> => {
		return api.post<{ success: boolean; message: string }>(`/notification-channels/${id}/test`);
	},

	testConfig: (data: {
		provider: string;
		config: Record<string, unknown>;
	}): Promise<{ success: boolean; message: string }> => {
		return api.post<{ success: boolean; message: string }>('/notification-channels/test', data);
	}
};

File: api/app/api/v1/notification_channels.py
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import CurrentSuperuser
from app.core.database import get_session
from app.services.notification_channel import NotificationChannelService
from shared.models.notification_channel import (
    NotificationChannelCreate,
    NotificationChannelRead,
    NotificationChannelTestConfig,
    NotificationChannelTestResult,
    NotificationChannelUpdate,
)

router = APIRouter(
    prefix="/notification-channels",
    tags=["notification-channels"],
)


def get_service(
    session: Annotated[AsyncSession, Depends(get_session)],
) -> NotificationChannelService:
    return NotificationChannelService(session)


@router.get("", response_model=list[NotificationChannelRead])
async def list_channels(
    _current_user: CurrentSuperuser,
    service: Annotated[NotificationChannelService, Depends(get_service)],
):
    return await service.list()


@router.post(
    "", response_model=NotificationChannelRead, status_code=status.HTTP_201_CREATED
)
async def create_channel(
    data: NotificationChannelCreate,
    current_user: CurrentSuperuser,
    service: Annotated[NotificationChannelService, Depends(get_service)],
):
    return await service.create(data, created_by=current_user.id)


@router.post("/test", response_model=NotificationChannelTestResult)
async def test_channel_config(
    data: NotificationChannelTestConfig,
    _current_user: CurrentSuperuser,
    service: Annotated[NotificationChannelService, Depends(get_service)],
):
    return await service.test_config(data.provider, data.config)


@router.get("/{id}", response_model=NotificationChannelRead)
async def get_channel(
    id: UUID,
    _current_user: CurrentSuperuser,
    service: Annotated[NotificationChannelService, Depends(get_service)],
):
    return await service.get(id)


@router.patch("/{id}", response_model=NotificationChannelRead)
async def update_channel(
    id: UUID,
    data: NotificationChannelUpdate,
    _current_user: CurrentSuperuser,
    service: Annotated[NotificationChannelService, Depends(get_service)],
):
    return await service.update(id, data)


@router.delete("/{id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_channel(
    id: UUID,
    _current_user: CurrentSuperuser,
    service: Annotated[NotificationChannelService, Depends(get_service)],
):
    await service.delete(id)


@router.post("/{id}/test", response_model=NotificationChannelTestResult)
async def test_channel(
    id: UUID,
    _current_user: CurrentSuperuser,
    service: Annotated[NotificationChannelService, Depends(get_service)],
):
    return await service.test(id)
